TypeScript project references and Nx Cloud

adriencaccia opened the pull request:
- Implement a correct usage of TypeScript project references using this article: https://moonrepo.dev/docs/guides/javascript/typescript-project-refs
- Setup Nx Cloud

adriencaccia commented:
@fargito WDYT?

fargito commented on nx.json:
```diff
@@ -14,7 +14,8 @@
         ],
-        "cacheDirectory": "nx-cache"
+        "cacheDirectory": "nx-cache",
+        "accessToken": "MzY4MGJjNWEtZmE5YS00ZWQ2LTgzOTEtYzY0Y2MzOTFlMGJkfHJlYWQtd3JpdGU="
```
@adriencaccia is this a real one?

adriencaccia replied:
It is yes!

fargito commented on tsconfig.json:
```diff
@@ -1,22 +1,18 @@
-  }
+  "extends": "./tsconfig.options.json",
+  "files": [],
+  "references": [
+    { "path": "./backend/orchestrator/tsconfig.json" },
+    { "path": "./contracts/orchestrator-contracts/tsconfig.build.json" },
+    { "path": "./packages/create-swarmion-app/tsconfig.json" },
+    { "path": "./packages/eslint-plugin/tsconfig.build.json" },
+    { "path": "./packages/nx-plugin/tsconfig.build.json" },
+    { "path": "./packages/serverless-cdk-plugin/tsconfig.build.json" },
+    { "path": "./packages/serverless-configuration/tsconfig.build.json" },
+    { "path": "./packages/serverless-contracts/tsconfig.build.json" },
+    { "path": "./packages/serverless-contracts-plugin/tsconfig.build.json" },
+    { "path": "./packages/serverless-helpers/tsconfig.build.json" },
+    { "path": "./user-docs/cloudfront/tsconfig.json" },
+    { "path": "./user-docs/documentation/tsconfig.json" }
```
This means we have to update this file when generating a new package/service right?

adriencaccia replied:
Yes, that way we can run [`tsc --build --verbose` at the root-level](https://moonrepo.dev/docs/guides/javascript/typescript-project-refs#on-all-projects)

fargito replied:
This needs to be populated in generators right?

adriencaccia replied:
I upgrade them!

fargito commented on contracts/orchestrator-contracts/tsconfig.json:
```diff
@@ -1,10 +1,13 @@
 {
-  "extends": "../../tsconfig.json",
+  "extends": "../../tsconfig.options.json",
```
Shouldn't this change be in `tsconfig.build.json` instead?

adriencaccia replied:
I followed [the guide](https://moonrepo.dev/docs/guides/javascript/typescript-project-refs#on-all-projects), and everything is working as expected

fargito reviewed (changes requested):
That would be cleaner indeed!

However, I'm confused to whether this change must apply only for `tsconfig.json`, `tsconfig.build.json` or both in libraries.

Also, we need to update generators as well to reflect this change.

fargito commented on .gitignore:
```diff
@@ -43,3 +43,6 @@ event.json
 nx-cache
+
+# Build cache manifests
+*.tsbuildinfo
```
I'm also wondering whether this should be cached by Nx as well. In this case, it should be included in the root Nx config

adriencaccia replied:
You are completely right 😉

adriencaccia commented:
> That would be cleaner indeed!
> 
> However, I'm confused to whether this change must apply only for `tsconfig.json`, `tsconfig.build.json` or both in libraries.
> 
> Also, we need to update generators as well to reflect this change.

To have a better understanding of the impacts, I would like to test on a bigger repo with known issues (@CorentinDoue I am looking at yours 👀). I will try to migrate it the same way I did in this PR and check if it solves some issues.

guillaumeduboc commented on packages/create-swarmion-app/package.json:
```diff
@@ -35,7 +35,7 @@
     "test-circular": "pnpm depcruise --validate .dependency-cruiser.js .",
     "test-linter": "pnpm linter-base-config .",
-    "test-type": "tsc --noEmit --emitDeclarationOnly false",
+    "test-type": "tsc",
```
why don't we do this modification in every package ?

adriencaccia replied:
To make `tsc --build` work in both packages and apps, we need to differentiate.
In apps, we will have `"test-type": "tsc",` and `"test-type": "tsc --noEmit --emitDeclarationOnly false",` in packages.

guillaumeduboc commented on contracts/orchestrator-contracts/tsconfig.json:
```diff
@@ -1,10 +1,13 @@
-    "outDir": "./dist/types"
+    "baseUrl": "src"
   },
+  "references": [
+    {
+      "path": "../../packages/serverless-contracts/tsconfig.build.json"
+    }
+  ],
```
is this still needed ? 
it is already specified in the root `tsconfig.options.json`

adriencaccia commented:
@fargito It's done 🎉 
I updated the generators and modified the examples.
I didn't have time to test on @CorentinDoue's project.
But I think we can merge anyway, the cleaning of the configs in itself is a good step forward.
And there are other cool things about PR 😉 
WDYT?

fargito reviewed (commented):
LGTM, thanks a lot for the hard work! :rocket:
